### PR TITLE
GEODE-6609: Protect from MetricsPublishingService exceptions

### PR DESCRIPTION
**Pull request by kirklund**
Catch and log Errors and RuntimeExceptions thrown from instantiation
of or calls to MetricsPublishingService implementations.

Co-authored-by: Michael Oleske <moleske@pivotal.io>

Please review: @demery-pivotal @mhansonp

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/metrics/CacheLifecycleMetricsSession.java [hunk 1]
```diff
@@ -19,20 +19,26 @@
```
Comment (demery-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/metrics/CacheLifecycleMetricsSession.java]: We also need to protect the call to `addSubregistry()`. I'm not sure about `removeSubregistry()` (I don't know if that ever throws).

File: geode-core/src/main/java/org/apache/geode/internal/metrics/CacheLifecycleMetricsSession.java [hunk 2]
```diff
@@ -68,7 +88,14 @@ public void cacheClosed(InternalCache cache) {
...
 
     for (MetricsPublishingService metricsPublishingService : metricsPublishingServices) {
-      metricsPublishingService.stop();
+      try {
+        metricsPublishingService.stop();
+      } catch (VirtualMachineError e) {
+        SystemFailure.initiateFailure(e);
+        throw e;
+      } catch (Error | RuntimeException e) {
+        logError(errorLogger, metricsPublishingService.getClass().getName(), e);
```
Comment (demery-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/metrics/CacheLifecycleMetricsSession.java]: This logs "Error invoking start," even though we invoked `stop()`. Perhaps pass the method name to `logError()`?

Comment (kirklund) [reply geode-core/src/main/java/org/apache/geode/internal/metrics/CacheLifecycleMetricsSession.java]: Fixed

File: geode-core/src/main/java/org/apache/geode/internal/metrics/CacheLifecycleMetricsSession.java (hunk 1 above)
Comment (kirklund) [reply geode-core/src/main/java/org/apache/geode/internal/metrics/CacheLifecycleMetricsSession.java]: The User would be invoking `MetricsSession.addRegistry` from `MetricsPublishingService.start(MetricsSession)`. For example, if the User tries to add a `CompositeMeterRegistry` then `CompositeMeterRegistry.add(MeterRegistry)` would throw IllegalArgumentException which our try-catch in `CacheLifecycleMetricsSession.cacheCreated(InternalCache)` would already be catching and logging.

Comment (demery-pivotal) [reply geode-core/src/main/java/org/apache/geode/internal/metrics/CacheLifecycleMetricsSession.java]: On `start(session)`, the service might simply stash the session, and defer adding the subregistry until a later time. In that situation, the `addSubregistry()` call can happen outside the safe context of `start(session)`. The risk is low… Is it low enough?

Comment (kirklund) [reply geode-core/src/main/java/org/apache/geode/internal/metrics/CacheLifecycleMetricsSession.java]: I think the user would have to use their own thread later on to invoke addSubregistry. Throwing to them is probably good enough. If we catch and log it in calls to addSubregistry, then I think we would want to rethrow so they get a Throwable.

**Pull request merged**